include/DoocsAdapter.h

```diff
@@ -1,36 +1,92 @@
-#ifndef _CHIMERATK_DOOCS_ADAPTER_H_
-#define _CHIMERATK_DOOCS_ADAPTER_H_
+// SPDX-FileCopyrightText: Deutsches Elektronen-Synchrotron DESY, MSK, ChimeraTK Project <chimeratk-support@desy.de>
+// SPDX-License-Identifier: LGPL-3.0-or-later
+#pragma once
 
 #include "CSAdapterEqFct.h"
 #include "DoocsPVFactory.h"
 
+#include <ChimeraTK/DataConsistencyGroup.h>
+#include <ChimeraTK/OneDRegisterAccessor.h>
+#include <ChimeraTK/ScalarRegisterAccessor.h>
+
 namespace ChimeraTK {
 
   /**
    * Base class used for all properties.
+   * Handles data consistency, ZMQ send, and synchronization with other DOOCS buffers
+   * All derived classes also must derive from DOOCS D_fct implementation.
    */
   class PropertyBase : public boost::enable_shared_from_this<PropertyBase> {
    public:
-    // List of other properties which need to update their DOOCS buffers when this property is written from the DOOCS
-    // side. This is used to synchronise multi-mapped PVs.
-    std::set<boost::shared_ptr<PropertyBase>> otherPropertiesToUpdate;
+    PropertyBase(EqFct* eqFct, std::string doocsPropertyName, DoocsUpdater& updater)
+    : _eqFct(eqFct), _doocsPropertyName(std::move(doocsPropertyName)), _doocsUpdater(updater) {}
+    virtual ~PropertyBase() = default;
+
+    /// returns associated DOOCS location
+    EqFct* getEqFct() { return _eqFct; }
+    /// returns associated DOOCS property. Not null.
+    D_fct* getDfct() { return dynamic_cast<D_fct*>(this); }
+    /// turns on ZeroMQ publishing
+    virtual void publishZeroMQ() { _publishZMQ = true; }
+    void setMacroPulseNumberSource(boost::shared_ptr<ChimeraTK::NDRegisterAccessor<int64_t>> macroPulseNumberSource);
+    void setMatchingMode(DataConsistencyGroup::MatchingMode newMode) { _consistencyGroup.setMatchingMode(newMode); }
 
-    // Update DOOCS buffer from PVs. The given transferElementId shall be used only for checking consistency with the
-    // DataConsistencyGroup. {} will be passed if the update is coming from another property (hence the update will
-    // only be taken with data matching set to none, which is the expected behaviour).
+   protected:
+    /// Update DOOCS buffer from PVs. The given transferElementId shall be used only for checking consistency with the
+    /// DataConsistencyGroup. {} will be passed if the update is coming from another property (hence the update will
+    /// only be taken with data matching set to none, which is the expected behaviour).
     virtual void updateDoocsBuffer(const TransferElementID& transferElementId) = 0;
 
-    virtual EqFct* getEqFct() = 0;
+    /// should be called for output vars mapped to doocs
+    /// registers processVar in data consistency group, initializes DOOCS error state and keeps a reference as _mainOutputVar
+    template<typename T>
+    void init(boost::shared_ptr<typename ChimeraTK::NDRegisterAccessor<T>> const& processVar);
```

Comment:
I did not copy this to the new style guide because it is not well defined and mostly referring to a trigger word (function called "init"). Rename it and it would no longer violate that point in the style guide. Maybe renaming is still worth considering: This function does not initialise every aspect of the object, it merely associates the property with the main output ProcessArray. So maybe the function should be called `associateMainOutputPV` or so?

Reply:
renamed init -> setupOutputVar